File: C++/Binary-Tree/Serialize and Deserialize Binary Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Codec {
public:

   // Encodes a tree to a single string.
    string serialize(TreeNode* root) {
        if (!root) return "#,";
        return to_string(root->val) + "," + serialize(root->left) + serialize(root->right);
    }

    // Decodes your encoded data to tree.
    TreeNode* deserialize(string data) {
        queue<string> q = convert(data); 
        return util(q);
    }
    
    TreeNode* util(queue<string>& q)
    {
        if (q.front() == "#") {
            q.pop();
            return NULL;
        } else {
            
            string cur = q.front();
            q.pop();
            
            TreeNode* root = new TreeNode(stoi(cur));
            
            root->left = util(q);
            root->right = util(q);
            
            return root;
        }
        
        
    }
    
    queue<string> convert(string& data)
    {
        queue<string> q;
        string cur = "";
        for (int i=0;i<data.size();i++)
        {
            if (data[i] == ','){
                q.push(cur);
                cur = "";
            } else {
                cur.push_back(data[i]);
            }
        }
        
        return q;
    }
    
};

// Your Codec object will be instantiated and called as such:
// Codec codec;
// codec.deserialize(codec.serialize(root));

File: Fb_ph/Practice/Python/Serialize and Deserialize Binary Tree.py
'''
Serialization is the process of converting a data structure or object into a sequence of bits so that it can be stored 
in a file or memory buffer, 
or transmitted across a network connection link to be reconstructed later in the same or another computer environment.

Design an algorithm to serialize and deserialize a binary tree. 
There is no restriction on how your serialization/deserialization algorithm should work. 
You just need to ensure that a binary tree can be serialized to a string and this string can be deserialized to the original tree structure.
'''

'''
Complexity
Time complexity: O(n)
Space complexity: O(n)
'''
'''
DFS
'''
class Codec:
    def serialize(self, root):
        if not root: return "#"
        return str(root.val) + "," + self.serialize(root.left) + "," + self.serialize(root.right)

    def deserialize(self, data):
        def deserialize_helper(queue):
            val = queue.popleft()
            if val == "#":
                return None
            node = TreeNode(int(val))
            node.left = deserialize_helper(queue)
            node.right = deserialize_helper(queue)
            return node
        
        queue = deque(data.split(","))
        return deserialize_helper(queue)
'''
BFS
'''
class Codec:
    def serialize(self, root):
        if not root:
            return ""
        queue = [root]
        result = []
        while queue:
            node = queue.pop(0)
            if node:
                result.append(str(node.val))
                queue.append(node.left)
                queue.append(node.right)
            else:
                result.append("#")
        return ",".join(result)

    def deserialize(self, data):
        if not data:
            return None
        values = data.split(",")
        root = TreeNode(int(values[0]))
        queue = [root]
        i = 1
        while queue:
            node = queue.pop(0)
            if values[i] != "#":
                left = TreeNode(int(values[i]))
                node.left = left
                queue.append(left)
            i += 1
            if values[i] != "#":
                right = TreeNode(int(values[i]))
                node.right = right
                queue.append(right)
            i += 1
        return root
